File: apps/api/app/modules/gate/ladder.py
"""The approval ladder.

The original brief asked for "gate entry approval under 2 seconds". That is physically
impossible — the round trip includes a human being picking up a phone and tapping a
button. So the target is split, and the case everybody actually complains about (the
resident does not answer) is a designed outcome rather than an edge case:

    t=0s   guard submits            → server ack, p95 < 800ms
    t=0s   high-priority push to every device on the unit
    t=20s  no answer                → IVR call + SMS
    t=45s  no answer                → apply the unit's standing rule
    t=90s  still nothing            → escalate to the on-duty committee contact

Each rung is a Cloud Tasks message scheduled at creation time and cancelled if the
resident responds first. Every rung that fires is written to `approval_rungs` and shown
back to the resident, so "I never got the notification" becomes checkable.

Without Google Cloud credentials the scheduling is logged instead of enqueued, so the
whole flow is exercisable locally.
"""

from __future__ import annotations

import datetime as dt
import uuid
from dataclasses import dataclass

import structlog
from sqlalchemy import select
from sqlalchemy.ext.asyncio import AsyncSession

from app.common.config import get_settings
from app.modules.gate.models import (
    Approval,
    ApprovalRung,
    ApprovalState,
    Rung,
    StandingAction,
    StandingRule,
    VisitorCategory,
)

log = structlog.get_logger(__name__)


@dataclass(frozen=True)
class LadderStep:
    rung: Rung
    delay_seconds: int


#: The ladder. Order and timings are product decisions, not tuning knobs — changing
#: them changes what a resident experiences at their gate.
LADDER: tuple[LadderStep, ...] = (
    LadderStep(Rung.push, 0),
    LadderStep(Rung.ivr, 20),
    LadderStep(Rung.standing_rule, 45),
    LadderStep(Rung.mc_escalation, 90),
)

#: Categories a standing rule may auto-approve. A guest is never auto-approved — an
#: unknown person at the gate always needs a human decision.
AUTO_APPROVABLE = frozenset(
    {VisitorCategory.delivery, VisitorCategory.courier, VisitorCategory.cab}
)


async def schedule_ladder(approval_id: uuid.UUID, society_id: uuid.UUID) -> None:
    """Enqueue one Cloud Task per rung, each at its own delay.

    Scheduled up front rather than chained, so a worker crash cannot strand a visitor
    at the gate: the later rungs are already queued independently.
    """
    settings = get_settings()

    if settings.tasks_are_stubbed:
        for step in LADDER[1:]:
            log.warning(
                "ladder_stub_scheduled",
                approval_id=str(approval_id),
                rung=str(step.rung),
                delay_seconds=step.delay_seconds,
                detail="Cloud Tasks not configured — rung logged, not enqueued.",
            )
        return

    from google.cloud import tasks_v2  # imported lazily: absent in local installs

    client = tasks_v2.CloudTasksAsyncClient()
    parent = client.queue_path(
        settings.gcp_project_id, settings.gcp_region, settings.cloud_tasks_queue
    )

    for step in LADDER[1:]:
        schedule_time = dt.datetime.now(dt.UTC) + dt.timedelta(seconds=step.delay_seconds)
        await client.create_task(
            parent=parent,
            task={
                "name": f"{parent}/tasks/{approval_id}-{step.rung}",  # idempotent enqueue
                "schedule_time": schedule_time,
                "http_request": {
                    "http_method": tasks_v2.HttpMethod.POST,
                    "url": f"{settings.worker_base_url}/jobs/approval-rung",
                    "headers": {"Content-Type": "application/json"},
                    "body": (
                        f'{{"approval_id":"{approval_id}",'
                        f'"society_id":"{society_id}",'
                        f'"rung":"{step.rung}"}}'
                    ).encode(),
                    "oidc_token": {"service_account_email": settings.worker_service_account},
                },
            },
        )

    log.info("ladder_scheduled", approval_id=str(approval_id))


async def record_rung(
    session: AsyncSession,
    society_id: uuid.UUID,
    approval_id: uuid.UUID,
    rung: Rung,
    *,
    result: str,
) -> None:
    session.add(
        ApprovalRung(
            id=uuid.uuid4(),
            society_id=society_id,
            approval_id=approval_id,
            rung=rung,
            fired_at=dt.datetime.now(dt.UTC),
            channel_result=result,
        )
    )


async def resolve_standing_rule(
    session: AsyncSession,
    society_id: uuid.UUID,
    approval: Approval,
) -> StandingAction:
    """Decide what happens at t=45s when nobody has answered.

    Most specific rule wins: a matcher on the visitor name beats a category rule, which
    beats the unit default. With no rule at all the visitor is asked to wait — never
    auto-approved, because silence is not consent.
    """
    rules = await session.scalars(
        select(StandingRule).where(
            StandingRule.society_id == society_id,
            StandingRule.unit_id == approval.unit_id,
            StandingRule.is_active.is_(True),
        )
    )

    best: StandingRule | None = None
    best_score = -1

    for rule in rules:
        score = 0
        if rule.matcher:
            if not approval.visitor_name:
                continue
            if rule.matcher.lower() not in approval.visitor_name.lower():
                continue
            score += 2
        if rule.category is not None:
            if rule.category is not approval.category:
                continue
            score += 1
        if score > best_score:
            best, best_score = rule, score

    if best is None:
        return StandingAction.ask_to_wait

    # A rule cannot auto-approve a category that requires a human decision, even if a
    # resident configured it that way — an unknown guest is not a parcel.
    if best.action is StandingAction.auto_approve and approval.category not in AUTO_APPROVABLE:
        log.info(
            "standing_rule_downgraded",
            approval_id=str(approval.id),
            category=str(approval.category),
            detail="auto_approve not permitted for this visitor category",
        )
        return StandingAction.ask_to_wait

    return best.action


async def apply_rung(
    session: AsyncSession,
    society_id: uuid.UUID,
    approval_id: uuid.UUID,
    rung: Rung,
) -> Approval | None:
    """Execute one ladder rung. Called by the worker when a Cloud Task fires.

    Returns None when the approval was already resolved — the normal case, since most
    residents answer the first push and the later tasks still fire.
    """
    approval = await session.get(Approval, approval_id)
    if approval is None or approval.society_id != society_id:
        return None

    if approval.state is not ApprovalState.pending:
        log.info("ladder_rung_skipped", approval_id=str(approval_id), rung=str(rung))
        return None

    now = dt.datetime.now(dt.UTC)

    if rung is Rung.ivr:
        from app.modules.notify.voice import place_approval_call

        result = await place_approval_call(society_id, approval)
        await record_rung(session, society_id, approval_id, Rung.ivr, result=result)

    elif rung is Rung.standing_rule:
        action = await resolve_standing_rule(session, society_id, approval)
        await record_rung(
            session, society_id, approval_id, Rung.standing_rule, result=str(action)
        )

        if action is StandingAction.auto_approve:
            approval.state = ApprovalState.auto_approved
            approval.resolved_at = now
            approval.resolution_rung = Rung.standing_rule
        elif action is StandingAction.deny:
            approval.state = ApprovalState.denied
            approval.resolved_at = now
            approval.resolution_rung = Rung.standing_rule
        # ask_to_wait leaves it pending so the committee rung can still fire.

    elif rung is Rung.mc_escalation:
        await record_rung(
            session, society_id, approval_id, Rung.mc_escalation, result="notified committee"
        )
        if approval.state is ApprovalState.pending:
            approval.state = ApprovalState.escalated

    return approval


async def resolve(
    session: AsyncSession,
    society_id: uuid.UUID,
    approval_id: uuid.UUID,
    *,
    approved: bool,
    resolved_by: uuid.UUID,
) -> Approval | None:
    """Resident taps approve or deny.

    First response wins. A late tap on an already-escalated request is accepted as a
    correction, but the rung that resolved it is recorded either way.
    """
    approval = await session.get(Approval, approval_id)
    if approval is None or approval.society_id != society_id:
        return None

    if approval.state in (ApprovalState.approved, ApprovalState.denied):
        return approval  # idempotent — a double tap is not an error

    approval.state = ApprovalState.approved if approved else ApprovalState.denied
    approval.resolved_at = dt.datetime.now(dt.UTC)
    approval.resolved_by = resolved_by
    if approval.resolution_rung is None:
        approval.resolution_rung = Rung.push

    log.info("approval_resolved", approval_id=str(approval_id), state=str(approval.state))
    return approval

File: apps/api/src/modules/gate/ladder.ts
/**
 * The approval ladder.
 *
 * The brief asked for "gate approval in under 2 seconds". That is not achievable and
 * pretending otherwise designs the wrong system: the round trip includes a human being
 * finding their phone and tapping a button. What we can guarantee is our own leg of it
 * — the server acknowledges in under 800 ms and the push lands in under 3 s — and then
 * handle the case that actually decides whether a society keeps the product.
 *
 * That case is **nobody answers**. It is the single most common complaint against the
 * incumbents, and here it is a designed path rather than an edge case:
 *
 *     t=0s   push to every device on the unit
 *     t=20s  no answer  → IVR call + SMS to the primary resident
 *     t=45s  no answer  → apply the unit's standing rule
 *     t=90s  no answer  → escalate to the on-duty committee member
 *
 * Every rung fired is recorded, so "I never got a notification" is answerable with
 * evidence instead of an apology.
 *
 * This module is deliberately pure — no database, no clock, no network. The timing
 * rules are the part that must be exactly right, and pure functions let the tests say
 * so without waiting 90 real seconds.
 */

export type Rung = "push" | "ivr" | "sms" | "standing_rule" | "mc_escalation";
export type StandingAction = "auto_approve" | "ask_to_wait" | "deny";
export type VisitorCategory =
  | "guest"
  | "delivery"
  | "cab"
  | "courier"
  | "service"
  | "staff";

export interface LadderStep {
  readonly rung: Rung;
  readonly atSeconds: number;
  readonly description: string;
}

/**
 * The ladder itself.
 *
 * The gaps are chosen against how people actually behave, not round numbers. 20 s is
 * about as long as a guard will stand there before phoning the flat himself — past that
 * he stops trusting the app and reverts to the intercom, and once he does that the
 * product is dead in that society. 45 s is long enough that a resident who was reaching
 * for their phone still gets to decide for themselves. 90 s means the visitor is not
 * left standing indefinitely while nobody takes responsibility.
 */
export const LADDER: readonly LadderStep[] = [
  { rung: "push", atSeconds: 0, description: "Push to every device on the unit" },
  { rung: "ivr", atSeconds: 20, description: "IVR call and SMS to the primary resident" },
  { rung: "standing_rule", atSeconds: 45, description: "Apply the unit's standing rule" },
  { rung: "mc_escalation", atSeconds: 90, description: "Escalate to the on-duty committee member" },
];

/** Server-side acknowledgement budget. Breaching this is an incident, not a slow day. */
export const ACK_BUDGET_MS = 800;

export interface StandingRule {
  readonly id: string;
  readonly category: VisitorCategory | null;
  /** Free-text match against the visitor or company name, e.g. "Amazon". */
  readonly matcher: string | null;
  readonly action: StandingAction;
  readonly isActive: boolean;
}

export interface VisitorRequest {
  readonly category: VisitorCategory;
  readonly visitorName: string | null;
}

/**
 * Which rung is due next, given what has already fired and how long we have waited.
 *
 * Returns `null` when the ladder is exhausted — the caller then leaves the approval in
 * `escalated` rather than inventing another step.
 *
 * Driven by elapsed time rather than by a timer chain so that a worker restart, a
 * delayed Cloud Task or a duplicate delivery all converge on the same answer. Recovery
 * is just "ask again", which is the property that makes this survivable in production.
 */
export function nextRung(fired: readonly Rung[], elapsedSeconds: number): Rung | null {
  const done = new Set(fired);
  for (const step of LADDER) {
    if (done.has(step.rung)) continue;
    if (elapsedSeconds >= step.atSeconds) return step.rung;
    // Steps are in ascending time order, so the first future step ends the search.
    return null;
  }
  return null;
}

/** When the next unfired rung becomes due, in seconds from the request. Null if none. */
export function secondsUntilNextRung(fired: readonly Rung[]): number | null {
  const done = new Set(fired);
  const pending = LADDER.find((s) => !done.has(s.rung));
  return pending ? pending.atSeconds : null;
}

/**
 * Specificity of a rule. Higher wins.
 *
 * A rule naming a company ("Amazon") beats one naming a whole category ("delivery"),
 * which beats a catch-all. This is what lets a resident say "let Amazon in, but ask me
 * about other deliveries" and have it mean what they expect.
 */
function specificity(rule: StandingRule): number {
  return (rule.matcher ? 2 : 0) + (rule.category ? 1 : 0);
}

/** Safety ordering for ties. Denying is always the safe side of a coin flip. */
const ACTION_CAUTION: Record<StandingAction, number> = {
  deny: 2,
  ask_to_wait: 1,
  auto_approve: 0,
};

/**
 * Pick the standing rule that applies, or `null` if none does.
 *
 * Two properties matter more than the matching itself:
 *
 * 1. **Most specific wins**, so a company-level rule overrides a category-level one.
 * 2. **Ties resolve toward caution.** If a unit somehow holds two equally specific
 *    rules that disagree — easy to produce by adding a rule twice from two devices —
 *    the more cautious action wins. A duplicate rule must never be the reason a
 *    stranger is let in, and the resident can always approve manually. Silently
 *    granting access on a coin flip is the one outcome with no recovery.
 */
export function matchStandingRule(
  rules: readonly StandingRule[],
  request: VisitorRequest,
): StandingRule | null {
  const name = request.visitorName?.trim().toLowerCase() ?? "";

  const applicable = rules.filter((rule) => {
    if (!rule.isActive) return false;
    if (rule.category && rule.category !== request.category) return false;
    if (rule.matcher) {
      const needle = rule.matcher.trim().toLowerCase();
      // Substring, because the guard types "Amazon delivery" and the rule says "Amazon".
      if (needle === "" || !name.includes(needle)) return false;
    }
    return true;
  });

  if (applicable.length === 0) return null;

  return applicable.reduce((best, rule) => {
    const bySpecificity = specificity(rule) - specificity(best);
    if (bySpecificity !== 0) return bySpecificity > 0 ? rule : best;

    const byCaution = ACTION_CAUTION[rule.action] - ACTION_CAUTION[best.action];
    if (byCaution !== 0) return byCaution > 0 ? rule : best;

    // Fully tied: order by id so the outcome is stable across processes and replays.
    return rule.id < best.id ? rule : best;
  });
}

/**
 * What the standing rule rung resolves the approval to.
 *
 * `ask_to_wait` deliberately does NOT resolve the approval — the visitor waits and the
 * ladder continues to committee escalation. It means "I have not decided", not "no".
 */
export function outcomeOfStandingRule(
  action: StandingAction,
): { resolved: true; state: "auto_approved" | "denied" } | { resolved: false } {
  switch (action) {
    case "auto_approve":
      return { resolved: true, state: "auto_approved" };
    case "deny":
      return { resolved: true, state: "denied" };
    case "ask_to_wait":
      return { resolved: false };
  }
}

/**
 * Clock drift between a guard handset and the server, in seconds.
 *
 * Guard device clocks are routinely wrong, sometimes by hours — they are cheap, shared,
 * and nobody owns keeping them right. Every piece of business logic uses `server_ts`;
 * `device_ts` is kept only so the audit trail can show both and the difference.
 *
 * Positive means the device is running behind the server.
 */
export function driftSeconds(deviceTs: Date, serverTs: Date): number {
  return Math.round((serverTs.getTime() - deviceTs.getTime()) / 1000);
}

/** Drift beyond this is worth surfacing to the committee — the handset needs attention. */
export const DRIFT_ALERT_SECONDS = 300;
